### ISSUE TITLE
Visability

### ISSUE THREAD

**Issue by JipSterk** [MichalLytek/type-graphql#796]
for a personal project i would like to expose a different schema for the current user and or their permissions.
I came across an [implementation](https://graphql-ruby.org/authorization/visibility.html#hiding-parts-of-the-schema) that would fit my needs.

i found some [psuedo-code](https://github.com/graphql/graphql-js/issues/113#issuecomment-663252167) and have some examples of how this feature would work

```typescript
@ObjectType()
export class Secret {
  @Field()
  id: String;

  @Field({ visible: (root, context) => context.user?.id === root.creatorId })
  mySecretField: String;
}


@Resolver()
export class SecretResolver {
  @Query({ visible: (context) => context.user?.permissions.includes("read:secret") })
  getSecret() {
    //...logic
  }

  @Mutation({ visible: (context) => context.req.headers.host === "https://admin.somedomain.com" })
  createSecret() {
    //...logic
  }
}

```

i might have some ideas on where to start implementing the psuedo-code but im not really familiar on how the schema generation works for `type-graphql` if you could show me a few pointers on where to start i would be happy to implement it

**Comment by MichalLytek** [comment]
> You’re developing a new feature and want to make a gradual release to only a few users first.

You can use authorization feature to guard access to some queries or fields:
https://typegraphql.com/docs/authorization.html

> You don’t want non-admin users to know about administration functions of the schema.

Looks like security by obscurity 😅 You can build separate schemas for public and private usage just by providing a different set of resolvers to build schema, an then routing users base on the auth level if they need to be exposed at the same port:

```ts
@Resolver()
class FirstResolver {
  @Query()
  firstQuery(): string {
    return "First Query";
  }
}
@Resolver()
class SecondResolver {
  @Query()
  secondQuery(): string {
    return "Second Query";
  }
}

const firstSchema = await buildSchema({
  resolvers: [FirstResolver],
});
const secondSchema = await buildSchema({
  resolvers: [SecondResolver],
});
```

Building schema is a really resource-heavy operation. Doing that on every request, based on `context.req.headers.host` or other dynamic and granular conditions is a horific idea and won't take place in TypeGraphQL.

**Comment by JipSterk** [comment]
First of all thanks for the reply.
I also gave a though about dynamically based resolvers based on `roles/permission` have not tried it yet but might want to check out 🤔.

Maybe if i explain my use case a bit more"
On the front end i use `@graphql/codegen` which uses introspection to generate client code. so maybe during development performance would take a hit because the introspection query will run all the `visible` checks. but im not familiar enough with the internals to know how much of a performance hit this would be.
Since i then can use the generated client and the selection of fields on the `query/mutation` doesn't change. Visibility checks doesn't really matter anymore since you couldn't change the selection of fields on the `query/mutation`

an other use case may be for me:
Also maybe as developer i would build an api using `graphql` on top of `type-graphql` and users who pay for my api get certain features through exposing new `fields/mutations/queries` on the schema

**Comment by MichalLytek** [comment]
If the visibility thing would be run only on the introspection query, then first, it's not doable in `graphql-js` - TypeGraphQL is a layer on top of that, not a standalone engine as GraphQL Ruby. Secondly, if it only affect the introspection, it means that the fields and queries still be available for the clients, same as disabling introspection in Apollo Server doesn't mean you can't send the query string written by hand. So you would still need the runtime logic of guarding the access to the fields and queries, by using auth feature or your own layer based on middlewares.

So this feature has nothing to do with security, only with the tooling. AFAIK you can't expose directives nor extensions via the introspection query. So what I would suggest is to put your metadata in the `extensions` of the queries and fields, then build a schema object, then write your own tool which will traverse all the types and fields of schema, strip of the queries/fields you don't want to show for the selected groups and then print that schema so all the tooling will be able to generate proper type definitions and not suggest the hidden part.

You can also put functions in the `extensions` that would have the more sophisticated logic there, so while traversing you can execute such functions. This way your solution can work with TypeGraphQL, Apollo, GraphQLJS, NestJS and Nexus 😉 

So to sum up: the introspection part is not possible and the runtime part would be complex and resource-heavy (building schema on demand). Currently available features allow to achieve similar goals, and we can use standard GraphQL JS capabilties, so I don't thing TypeGraphQL would have such feature built-in 🔒
